### PR TITLE
expression: make cast return error if cast binary literal to another character set

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

`cast` function may cast string cross the different charset, like `binary` to `utf8m4`. if the binary literal is invalid bytes, we will return error

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -285,6 +285,9 @@ func (c *castAsStringFunctionClass) getFunction(ctx sessionctx.Context, args []E
 	}
 	bf.tp = c.tp
 	if args[0].GetType().Hybrid() || IsBinaryLiteral(args[0]) {
+		ft := args[0].GetType().Clone()
```
Comment (wjhuang2016) [inline expression/builtin_cast.go:288]: Please add a comment for it.

Comment (xiongjiwei) [reply expression/builtin_cast.go:290]: updated.

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -285,6 +285,11 @@ func (c *castAsStringFunctionClass) getFunction(ctx sessionctx.Context, args []E
 	}
 	bf.tp = c.tp
 	if args[0].GetType().Hybrid() || IsBinaryLiteral(args[0]) {
+		// when cast from binary to some other charsets, we should check if the binary is valid or not.
```
Comment (wjhuang2016) [inline expression/builtin_cast.go:288]:
```suggestion
		// When cast from binary to some other charsets, we should check if the binary is valid or not.
```

**Review by wjhuang2016 — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**